Python 3

douglas-raillard-arm opened the pull request:
Timely dodge Python 2 deprecation (January 1st 2020) with a Python 3 conversion, and cleanups as Christmas presents.

The main changes are:
* Convert to Python 3
* Turn the project into a "regular" one that can be installed with pip in a venv, which will install the dependencies as required and provide a CLI stub in a standard location
* Cleanup some nasty code (mostly around uses of "except" and assert that are not functionally correct in some circumstances) and also some less important but basic PEP8 compliance.

The main issue I faced was updating tncc.py to use bytes instead of str to deal with binary data. That's a necessary but pretty invasive change, and I may have forgotten some spots since I'm not sure to achieve 100% coverage with my use case. Use of the wrong type will raise a TypeError in most cases, except when embedded inside other str/bytes: str(b'foo') == "b'foo'". This happens when building URLs and leads to silently building the wrong request.

DouglasRaillard committed:
Convert to Python 3 with 2to3

DouglasRaillard committed:
Turn into a distributable Python project

* Move juniper-vnp.py to junipervpn/vpn.py and tncc.py to
  junipervpn/tncc.py, thereby making them modules in the junipervpn
  package.

* Provide a setup.py script, suitable for use as "pip install ." with
  the necessary list of dependencies, and the CLI tool entry point.
  Installing this way will provide a "juniper-vpn.py" binary in a
  standard location in the PATH.

* Add compatibility shim for backward compatibility with the old
  installation method. It includes a warning as using that shim is not
  recommended at all.

chrisdiamand commented on juniper-vpn.py:
```diff
@@ -1,17 +1,16 @@
-#!/usr/bin/env python2.7
-# -*- coding: utf-8 -*-
+#!/usr/bin/env python3
```
This commit LGTM - and presumably this one is 100% automated?

douglas-raillard-arm replied:
Yep, that's the blanket 2to3, so its easy to re-roll.

chrisdiamand commented on junipervpn/vpn.py:
```diff
@@ -386,8 +386,10 @@ def main():
-    jvpn.run()
+    try:
+        jvpn.run()
+    finally:
```
Cool, this LGTM - I've just done a quick test and this does have the correct behaviour on Ctrl-C (i.e. stop *is* always called). This is a much nicer way to do it that the signal handler :)

chrisdiamand commented on junipervpn/tncc.py:
```diff
@@ -195,7 +195,7 @@ def encode_0cf1(s):
 def encode_0cf3(i):
     return encode_packet(0x0013, 1, struct.pack("<I", i))
 
-class x509cert(object):
```
Seems about right :)

chrisdiamand commented on junipervpn/vpn.py:
```diff
@@ -16,12 +15,15 @@
-import netifaces
 import datetime
 
+import mechanize
```
Same here

chrisdiamand commented on junipervpn/tncc.py:
```diff
@@ -14,16 +13,20 @@
 import pyasn1_modules.rfc2459
 import pyasn1.codec.der.decoder
-import xml.etree.ElementTree
+import mechanize
```
Is there a particular reason you've moved these imports out of the main group?

douglas-raillard-arm replied:
There is a widespread convention blessed by PEP8 around modules grouping:
> Imports should be grouped in the following order:
>    * Standard library imports.
>    * Related third party imports.
>    * Local application/library specific imports.

You should put a blank line between each group of imports.
https://www.python.org/dev/peps/pep-0008/#imports

The main reason behind that I guess is to be able to quickly visually jump over stdlib imports, since they are "uninteresting" and pretty harmless (they will always be available), the real meat being the project dependencies and the project-specific stuff.

chrisdiamand replied:
Makes sense, I didn't know about that :)

chrisdiamand commented on junipervpn/tncc.py:
```diff
@@ -320,7 +320,7 @@ def parse_response(self):
         for line in self.r.readlines():
             line = line.strip()
             # Note that msg is too long and gets wrapped, handle it special
```
All makes sense!

chrisdiamand commented on junipervpn/tncc.py:
```diff
@@ -121,7 +121,7 @@ def decode_0cf3(buf, indent):
 def decode_packet(buf, indent=""):
     cmd, _1, _2, length, _3 = struct.unpack(">IBBHI", buf[:12])
```
LGTM

chrisdiamand commented on junipervpn/vpn.py:
```diff
@@ -306,7 +306,6 @@ def stop(self):
                 self.child.wait()
             except OSError:
                 pass
-        sys.exit(0)
```
LGTM

chrisdiamand commented on junipervpn/vpn.py:
```diff
@@ -310,46 +311,45 @@ def stop(self):
-    parser.add_argument('-u', '--username', type=str,
+    parser.add_argument('-u', '--username',
                         help='User name')
-    parser.add_argument('-p', '--pass_prefix', type=str,
```
I'd be tempted to leave these as they are (i.e. with `type=str`) if that's OK - I guess I slightly prefer it because it means I don't have to remember what the default argument type is :)

chrisdiamand replied:
Thanks!

chrisdiamand commented on junipervpn/vpn.py:
```diff
@@ -310,46 +311,45 @@ def stop(self):
+    parser.add_argument('-d', '--device-id',
                         help="Hex device ID")
     parser.add_argument('-f', '--enable-funk', action='store_true',
+                        default=None,
```
OK - this has to be `None`, rather than `False`, because it's handled in a function which doesn't know if it's come from `argparse` or a config file (where it could be a string like `"no"`).

chrisdiamand commented on junipervpn/vpn.py:
```diff
@@ -310,46 +311,45 @@ def stop(self):
+    parser.add_argument('-U', '--user-agent',
                         help="User agent string")
     parser.add_argument('action', nargs=argparse.REMAINDER,
+                        default=[],
```
OK

chrisdiamand commented on junipervpn/vpn.py:
```diff
@@ -310,46 +311,45 @@ def stop(self):
-    if not len(args.action):
-        args.action = None
+    with contextlib.suppress(ValueError):
+        args.action.remove('--')
```
Is this correct? `--` can only be the very first argument; we don't want to remove subsequent ones, because they could be part of the command to be executed.

douglas-raillard-arm replied:
Makes sense. It would still be usable with remove() but maybe it's better to look for that in the first item only, I'll make the change:

'foo -- bar --' would indeed be turned into 'foo bar --', but the correct result can be achieved with '-- foo -- bar --'

chrisdiamand commented on junipervpn/vpn.py:
```diff
@@ -186,9 +186,10 @@ def action_tncc(self):
         dssignin_cookie = self.find_cookie('DSSIGNIN')
-        t = tncc.TNCC(self.args.host, args.device_id, args.enable_funk,
+        args = self.args
+        t = tncc.TNCC(args.host, args.device_id, args.enable_funk,
```
Good catch!

chrisdiamand commented on junipervpn/tncc.py:
```diff
@@ -78,9 +78,9 @@ def decode_0ce4(buf, indent):
 
 # 0ce5 - string without hex prefixer
 def decode_0ce5(buf, indent):
-    s = struct.unpack(str(len(buf)) + "s", buf)
```
This all seems OK - although I'm surprised this wasn't one of the first things in the series... does it still work with Python3 without this patch?

(I mean, we should definitely commit it, but it would be cool to ensure any commit within this PR will work).

douglas-raillard-arm replied:
> does it still work with Python3 without this patch ?

Not really, you can't mix Unicode strings and bytestring without explicit conversion in Python 3 (which is a great improvement, but makes conversion a tad painful). The main reason it's not the first thing in the series is that I worked my way through various fixes, and "everything was working except tncc so I'll keep that for the end".

chrisdiamand reviewed (changes requested):
Thanks for this!

I've put a few pretty minor comments on. Also, there's been a couple of "recent" (well, 2018...) changes in the original repo at `russdill/juniper-vpn-py` - I might go ahead and merge those before we go and rename all the files. If you wouldn't mind rebasing on top of that that would be great!

Cheers!

douglas-raillard-arm commented:
Hi @chrisdiamand, thanks for the review.

I've made the following changes:
* rebased on master
* added back type=str to ArgumentParser.add_argument()
* reshuffled the commits order to have the tncc bytestring update close to the beginning
* Fixed the `--` in args.action thing

Since I haven't tested that new version of the PR at all, I've made a backup of the current state of the branch at: https://github.com/douglas-raillard-arm/juniper-vpn-py/tree/python3.bak

chrisdiamand commented:
Thanks!

I've just tried it out, but it looks like the proper Python project layout is incompatible with how I was invoking it (i.e. running juniper-vpn.py directly). It would be neat to be able to use this without _having_ to run `setup.py` - presumably we could achieve this by just adding a wrapper script, `juniper-vpn.py` at the top level that calls into `vpn.main()`? I don't know if this would conflict/duplicate the entry point created by `setup.py` though.

More seriously though, it looks like there are no Debian packages for Python3 versions of `mechanize` or `urlgrabber`. `urlgrabber` actually seems like it's not being used, but I'm thinking that for now it might be sensible to cherry-pick as much as possible, then do the final switch to Python3 a bit later, in case one of us gets a chance to see if we can replace the `mechanize` usage with something that's more actively maintained (`mechanicalsoup` seems to at least have a debian package?).

douglas-raillard-arm commented:
Allowing to execute without installing can sound appealing, but that also breaks the invariants that the install process will ensure. In this case, it means you end up manually installing the dependencies, which might be missing from your distro.

It's possible to use `pip install . --no-dependencies --editable`, which should only create the wrapper script calling the entry point. Apparently, you can also do that if you want it to /usr/local/bin (not tested):
```
pip install . --no-dependencies --editable --install-option="--prefix=/usr/local/bin"
```
These kind of things are best left to distro package maintainers though. install-less scripts are usable for _very_ simple projects that only use the standard lib, but anything more complicated than that is much better served by pip.

chrisdiamand commented:
I've just cherry-picked all of the easy commits :) No objections to the others either really but I need a spare moment to do a bit more research!

The remaining commits are:
* `9cc919a junipervpn.vpn: Sanitize option handling`

Looks about right but contextlib.suppress was introduced in Python 3.4, so it
can't be committed until after switching to Python 3. Also I wonder if we can
roll the handling of `action` and `enable_funk` into the main loop that handles
getting all the other arguments from the config file somehow.

* `be370e4 junipervpn.vpn: Remove sys.exit(0) in stop()`

Need to think about this one - it seems correct but I think we might need some
special handling for Ctrl-C. At the moment it now generates *two* backtraces
(previously "only" one).

* `c7825c1 junipervpn.tncc: Use bytestring`

I found a missing one during testing:
```
@@@ -551,7 -551,7 +551,7 @@@ class TNCC
              if cert_id not in certs:
                  logging.warn('Could not find certificate for %s', str(req_dns))
  
--        inner = ''
++        inner = b''
```
* `5e61d3d Turn into a distributable Python project:`

I don't know enough about setuptools yet, I'll do some research when I get a
free moment :)

I am aware that quite a lot of people are using this now though, and I don't want to immediately go and break all of their workflows by requiring them to install it with `setup.py`. So I'm still quite keen on keeping a `juniper-vpn.py` script around at the top-level to minimise disruption.

* `8f3684f Convert to Python 3 with 2to3`

I'd like to have a go at removing the mechanize dependency before doing this
one (similar reasons to the above).

Also, I committed `3df670d junipervpn: Use camel case for class names` as-is - but I guess I should mention that effectively makes all the classes old-style on Python 2, but
nothing's doing any inheritance or type checking on these so it should be OK.

Cheers!

douglas-raillard-arm commented:
>Need to think about this one - it seems correct but I think we might need some
special handling for Ctrl-C. At the moment it now generates two backtraces
(previously "only" one).

Do you have an example ? (I'm currently having issues probably unrelated to this PR that prevent me from getting to that stage)

douglas-raillard-arm commented:
> I found a missing one during testing:
That's been fixed by @theotherjimmy , I'll update the branch soon

douglas-raillard-arm commented:
> So I'm still quite keen on keeping a juniper-vpn.py script around at the top-level to minimise disruption.

I'll add a juniper-vpn.py shim with a warning telling people they should not use this shim, and rely on the pip instead. I found that it's almost never a good idea to use the distro Python packages unless the application itself is also packaged by the distro. When taking something straight of github, it's a much better idea to use a venv/virtualenv.

> go and break all of their workflows by requiring them to install it with setup.py

Since their setup will be broken anyway by having to find and install Python3 packages for their distro, they would probably loose less time installing python3-venv and running the documented method rather than dealing with the dependencies manually

douglas-raillard-arm commented:
Hi @chrisdiamand,
I've updated the PR with the following changes:

* added shim as juniper-vpn.py with a warning telling users to install using pip
* a unified CLI and config file merge system so that all options are handled the same way
* a --verbose option to allow debugging HTTP requests without editing the `debug` global variable
* "clean" exit with ctrl-C: the backtrace is hidden and replaced by the message:
```User interrupt, stopping the VPN```

chrisdiamand commented on juniper-vpn.py:
```diff
@@ -347,6 +346,7 @@ def stop(self):
                         help='External command')
 
     args = parser.parse_args()
+    args.__dict__['username'] = None
```
This element in the namespace is already created by `add_argument` - so if you overwrite it here, users won't be able to set the username on the command line. Can you remove this line please?

douglas-raillard-arm replied:
good point, I dont really know what this line is doing there ...

chrisdiamand commented on junipervpn/vpn.py:
```diff
@@ -311,82 +312,116 @@ def stop(self):
+
+    # Tuple of (config getter, post process)
+    # for each CLI argument
+    arg_map = defaultdict(
```
This is kind of cool, but might be way overkill :) I wonder if another way might be to do two argument parsing passes?

* `parse_known_args()` with just the `-c` option, to find the config file
* Set up the "real" argument parser, using defaults from the config file if available
* Run `parse_args()`

But either way - would you mind splitting this out into a separate PR? I don't think it should block making Python 3 work. I wonder if we could also squash in the "Sanitize option handling" commit too.

douglas-raillard-arm replied:
As long as the command line can be partially parsed unambiguously that should be fine, I'll give that a try. That said, the will only spare the `Default()` instances, the logic to parse the config file should pretty much stay unchanged.

OTH dynamically adding arguments based on partially parsed command line can turn into a big headache:
https://github.com/ARM-software/lisa/blob/master/tools/exekall/exekall/main.py#L421

chrisdiamand replied:
Ouch - although that does use subparsers, _hopefully_ we don't need quite that much complexity :)

chrisdiamand commented on junipervpn/vpn.py:
```diff
@@ -311,7 +311,6 @@ def stop(self):
                 self.child.wait()
             except OSError:
                 pass
-        sys.exit(0)
```
Would you mind squashing this with the other exit behaviour commit please? And maybe splitting into a separate PR, as it's not directly related to the Python 3 support.

chrisdiamand commented on junipervpn/vpn.py:
```diff
@@ -25,8 +25,6 @@
 
 from junipervpn import tncc
 
-debug = False
```
Should we make the equivalent change in `tncc.py` too?

And again, maybe this could be split out?

chrisdiamand commented on junipervpn/tncc.py:
```diff
@@ -79,7 +79,7 @@ def decode_0ce4(buf, indent):
 def decode_0ce5(buf, indent):
     s = struct.unpack(str(len(buf)) + "s", buf)[0]
     logging.debug('%scmd 0ce5 (string) %d bytes', indent, len(buf))
-    s = s.rstrip('\0')
```
This all looks good to me, I'll go ahead and cherry-pick this commit onto `master` ASAP.

chrisdiamand commented on juniper-vpn.py:
```diff
@@ -0,0 +1,11 @@
+
+    # Import after the warning, so the warning is the first thing seen rather
+    # than some ImportError due to some missing packages
+    from junipervpn import vpn
```
I think `tncc.py` is meant to be usable standalone too - that's probably why currently both scripts are in the top level, rather than having `tncc.py` in its own module.

Would you mind adding the shim in the same commit as the directory rearranging please?

douglas-raillard-arm replied:
I'll add a shim and an entry point as well, since it's probably a legitimate use as well

chrisdiamand commented on juniper-vpn.py:
```diff
@@ -42,7 +41,7 @@
 
 def truncated_value(h):
-    bytes = map(ord, h)
+    bytes = list(map(ord, h))
```
I've merged the backwards-compatible parts of this commit.

I'm sort of thinking it would be nice to do all the preparatory patches first, then flip the switch to Python 3 at the end. That way individual commits within the series will still work for everyone.

That would include the setup.py stuff - I guess we can't break Python 2 support without that, because installing via `pip` is going to be the easiest way for most people to get `mechanize` on Python 3.

douglas-raillard-arm replied:
Writing code that works for both Python 2 and Python 3 is generically speaking a mess and requires the use of packages like `future` or `six`. Unless there is a compelling reason to do so, it's a much, much better idea to avoid getting into that. The situation usually is:
1) Python 2 is dead. Some distributions don't ship python2 by default. All distributions without exceptions that I know ship Python 3, and more specifically >= 3.4
2) the only really good reason to maintain Python 2 compatibility is to be able to mix with old/unmaintained code

Point 1) means that *all* users will have access to Python 3. At the very worst, it will require an `apt install python3`. Point 2) means we don't care about Python2, since this project is not a lib anyone else is trying to import, and this package does not rely on any 3rd party library that is Python 2 only.

chrisdiamand replied:
> this package does not rely on any 3rd party library that is Python 2 only.

This is only partially true, due to the mechanize dependency. `python3-mechanize` hasn't yet hit Debian stable. It can be installed from Pip, but so far I have managed to get away with having zero user-installed Python packages... hence my reluctance to start now :)

This is also why I appreciate adding the wrapper scripts :) Although I reckon I might end up deleting the warnings...

Re. a "flag day" switch - having thought about it, it probably is OK. I was worried about people suddenly not being able to get into the VPN, but that would only happen if they've just updated this repo for some reason, in which case it'll be pretty obvious what's gone wrong.

chrisdiamand reviewed (commented):
Thanks for the update!

chrisdiamand commented on junipervpn/tncc.py:
```diff
@@ -79,7 +79,7 @@ def decode_0ce4(buf, indent):
 def decode_0ce5(buf, indent):
```
I've merged this one now so I think it can be dropped from the PR :)

douglas-raillard-arm commented:
@chrisdiamand I've rebased and updated the PR. Another round of careful check would be a very good idea as I got a fairly high amount of merge conflict due to reordering of the bytestring commit that was cherry picked independently.

chrisdiamand commented on setup.py:
```diff
@@ -0,0 +1,41 @@
+        'pyasn1',
+        'pyasn1-modules',
+        'mechanize',
+        'urlgrabber',
```
Did you figure out what urlgrabber is used for? I removed the import statement for it, it doesn't seem to be actually used by any of the code directly at least.

chrisdiamand commented on README:
```diff
@@ -1,3 +1,17 @@
+Installation
+============
```
Can't see how to comment on the commit message - but I think there's a typo in "junpier-vpn" :)

chrisdiamand commented on README:
```diff
@@ -1,3 +1,17 @@
+
+$ python3 -m venv myvenv
+$ source myvenv/bin/activate
+$ pip install juniper-vpn-py
```
Testing this out locally, I get this:
```
Downloading https://files.pythonhosted.org/packages/ef/05/4b773f74f830a90a326b06f9b24e65506302ab049e825a3c0b60b1a6e26a/pycurl-7.43.0.5.tar.gz (216kB)
    100% |████████████████████████████████| 225kB 3.6MB/s 
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "/tmp/pip-install-al77vdua/pycurl/setup.py", line 234, in configure_unix
        stdout=subprocess.PIPE, stderr=subprocess.PIPE)
      File "/usr/lib/python3.7/subprocess.py", line 775, in __init__                                     
        restore_signals, start_new_session)
      File "/usr/lib/python3.7/subprocess.py", line 1522, in _execute_child                              
        raise child_exception_type(errno_num, err_msg, err_filename)                                     
    FileNotFoundError: [Errno 2] No such file or directory: 'curl-config': 'curl-config'                 

    During handling of the above exception, another exception occurred:                                  

    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-install-al77vdua/pycurl/setup.py", line 961, in <module>                            
        ext = get_extension(sys.argv, split_extension_source=split_extension_source)                     
      File "/tmp/pip-install-al77vdua/pycurl/setup.py", line 623, in get_extension                       
        ext_config = ExtensionConfiguration(argv)
      File "/tmp/pip-install-al77vdua/pycurl/setup.py", line 101, in __init__                            
        self.configure()
      File "/tmp/pip-install-al77vdua/pycurl/setup.py", line 238, in configure_unix                      
        raise ConfigurationError(msg)
    __main__.ConfigurationError: Could not run curl-config: [Errno 2] No such file or directory: 'curl-config': 'curl-config'
```
... installing `libcurl4-openssl-dev` fixes this, but then I get:
```
Running setup.py bdist_wheel for netifaces ... error                                           [52/1424]  Complete output from command /home/chris/bin/_bundle/juniper-vpn-py/myvenv/bin/python3 -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-6tl3r9g1/netifaces/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /tmp/pip-wheel-8r60e7v9 --python-tag cp37:                                                         
  usage: -c [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]                                        
     or: -c --help [cmd1 cmd2 ...]               
     or: -c --help-commands               
     or: -c cmd --help                                                                                    
                                                    
  error: invalid command 'bdist_wheel'    
                                                                                                          
  ----------------------------------------
(same again for netifaces, urlgrabber, pycurl)
```
Fixed by `pip3 install wheel`.

And finally:
```
Failed to build juniper-vpn netifaces urlgrabber pycurl                                         [156/1575]
Installing collected packages: six, webencodings, html5lib, mechanize, netifaces, pyasn1, pyasn1-modules,
pycurl, urlgrabber, juniper-vpn                                                                           
  Running setup.py install for netifaces ... done
  Running setup.py install for pycurl ... error    
    Complete output from command /home/chris/bin/_bundle/juniper-vpn-py/myvenv/bin/python3 -u -c "import s
etuptools, tokenize;__file__='/tmp/pip-install-6tl3r9g1/pycurl/setup.py';f=getattr(tokenize, 'open', open)
(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --
record /tmp/pip-record-1g4t0008/install-record.txt --single-version-externally-managed --compile --install
-headers /home/chris/bin/_bundle/juniper-vpn-py/myvenv/include/site/python3.7/pycurl:                     
    Using curl-config (libcurl 7.64.0)                                                                   
    Using SSL library: OpenSSL/LibreSSL/BoringSSL                                                         
    running install                      
    running build                                                                                        
    running build_py                                                                                     
    creating build                                                                                        
    creating build/lib.linux-x86_64-3.7                                                                   
    creating build/lib.linux-x86_64-3.7/curl                                                                  copying python/curl/__init__.py -> build/lib.linux-x86_64-3.7/curl       
    running build_ext                                                                                    
    building 'pycurl' extension                                   
    creating build/temp.linux-x86_64-3.7
    creating build/temp.linux-x86_64-3.7/src
    x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werr
or=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -DPYCURL_VERSION="7.43.0.5" -DHAVE_CURL_SSL=1 -DH
AVE_CURL_OPENSSL=1 -DHAVE_CURL_SSL=1 -I/home/chris/bin/_bundle/juniper-vpn-py/myvenv/include -I/usr/includ
e/python3.7m -c src/docstrings.c -o build/temp.linux-x86_64-3.7/src/docstrings.o
    In file included from src/docstrings.c:4:                                                            
    src/pycurl.h:165:13: fatal error: openssl/ssl.h: No such file or directory
     #   include <openssl/ssl.h>                                                                         
                 ^~~~~~~~~~~~~~~                                                                          
    compilation terminated.                                                                                   error: command 'x86_64-linux-gnu-gcc' failed with exit status 1 
```
Fixed by `apt install libssl-dev`.

Nothing to change in the PR really, I might add some extra documentation later though.

douglas-raillard-arm replied:
I've updated the README, thanks :)

chrisdiamand commented on README:
```diff
@@ -1,3 +1,17 @@
+$ pip install juniper-vpn-py
+
+Once that venv activated, the "juniper-vpn.py" binary will be in the path.
+NOTE: DO NOT USE ./juniper-vpn.py shim distributed in the git repository, use
```
Out of interest, does it work to install it without the virtualenv? I wonder if you can do something like this?
```
pip3 install git+https://github.com/chrisdiamand/juniper-vpn-py
```
If so, what's the benefit of the virtualenv? I'd guess that people who want to/know how to use them will use them anyway, but I expect most people to just want to install this in the fewest number of steps.

douglas-raillard-arm replied:
It does work, but that would do a system wide install. This means that pip will go and write in the site-packages directory, e.g. /usr/lib/python3.8/site-packages. The problem with that is that the exact same folder is used to install packages using the distro package manager like apt. 

That means that pip will override files managed by the distro package manager, potentially updating a package to a version that will break other installed software packaged by your distro. If you pass --user to pip, you will install in the HOME but otherwise the issue will remain the same as the user-local site-packages directory takes precedence (as seen in `sys.path`).

The only foolproof way of avoiding that is to use a venv/virtualenv so that the installation is completely isolated and will not conflict with anything else on the system. 

That means the following:
* system-wide `pip install` should only be used to create distro packages, or for someone ready to deal with the breakage that will follow its use (on some very up-to-date systems like Archlinux, it's not too much of an issue as the distro packages the latest version anyway). Also, system wide `pip install` will need to be ran as root although it might trigger complex work including building software.
* `pip install --user` is to be treated the same way, although the mess is easier to cleanup
* `venv` (or `virtualenv`) is the only trustworthy recommendation that can be done when it's supposed to be distribution-agnostic .

chrisdiamand replied:
OK, let's stick with that then. Just seems a bit odd to fire up a virtualenv just to connect to the VPN :)

douglas-raillard-arm replied:
Fortunately that's straightforward and easy to setup. I've come across insane setups based on wild use of PYTHONPATH, which is most of the time a recipe for endless pain ...

chrisdiamand commented on junipervpn/vpn.py:
```diff
@@ -0,0 +1,395 @@
+import datetime
+
+import mechanize
+import netifaces
```
Isn't this a duplicate import? We already get netifaces on line 19.

chrisdiamand commented on junipervpn/vpn.py:
```diff
@@ -0,0 +1,395 @@
+                        help='External command')
+
+    args = parser.parse_args()
+    args.password = None
```
Apart from a removed import __future__, changing `args.__dict__[...]` to `args.password` seems to be the only non-required change added during the rename of this file.

It's not super important, but maybe worth reverting in the interests of accurate history? (i.e. to keep this as close to a straight rename as possible). It could then be improved in the argparse PR.

chrisdiamand reviewed (commented):
I think I've come to terms with my philosophical objections so let's merge this :) I've just added a few last points.

douglas-raillard-arm commented:
> replace the mechanize usage with something that's more actively maintained

I had a quick look at the package and it is actively maintained, and have a package in Debian testing and unstable:
https://pypi.org/project/mechanize/#history
https://packages.debian.org/unstable/python3-mechanize
https://packages.debian.org/testing/python3-mechanize

One more reason to use the venv for cross-distro install procedure :)

chrisdiamand commented:
> I had a quick look at the package and it is actively maintained, and have a package in Debian testing and unstable:
> https://pypi.org/project/mechanize/#history
> https://packages.debian.org/unstable/python3-mechanize
> https://packages.debian.org/testing/python3-mechanize
> 
> One more reason to use the venv for cross-distro install procedure :)

I think when I wrote that comment in December I had just read a bunch of stackoverflow posts and similar talking about the hiatus ~2011-2017... it does look like a bit more has happened since then :)

I'd seen the debian package - that's one way I've been testing this PR :) Although it's also worth noting that the first Ubuntu version containing it will be the 20.04 release this month: https://packages.ubuntu.com/search?suite=default&section=all&arch=any&keywords=python3-mechanize&searchon=names

douglas-raillard-arm commented:
I guess they realized some Python 3 packages were missing from the distro repo as Python 2 died. AFAIR Ubuntu is based on Debian testing so that's not entirely surprising. I wish there was a straightforward tool like checkinstall to turn a Python package into a distro package (even if that package would not meet all the distro requirements like for Debian). Maybe there is ?

chrisdiamand commented:
Merged - thanks!
